docs/Dialects/Verif.md

```diff
@@ -4,6 +4,334 @@ This dialect provides a collection of operations to express various verification
 
 [TOC]
 
+## Contracts
+
+Formal contracts are a key building block of the Verif dialect to help make formal verification scale to larger designs and deep module hierarchies.
+Contracts describe what a circuit expects from its inputs (`verif.require`) and what it guarantees its output to be (`verif.ensure`).
+These contracts can then be used in two key ways:
+
+1.  A contract can be _verified_ to hold by turning `require`s into `assume`s and `ensure`s into `assert`s.
+    Doing so verifies that a circuit upholds the contract by placing asserts on the values it produces, and by placing assumes on the input values the circuit sees.
+    In a nutshell, this checks that, assuming the inputs honor the contract, the circuit also upholds the contract.
+    This check can be done very efficiently by creating `verif.formal` ops to verify each contract.
+
+2.  A contract can be _assumed_ to hold by turning `require`s into `assert`s and `ensure`s into `assume`s.
+    Doing so verifies that the inputs fed into a circuit uphold the contract, such that the outputs can be assumed to have the promised values.
+    In a nutshell, this checks that the inputs to a circuit honor the contract and therefore the circuit can be assumed to uphold the contract.
+    Assuming a contract can often eliminate large parts of the circuit's actual implementation, since the contracts tend to be a simpler description of a circuit's functionality.
+
+### Multiply-by-9 Example
+
+Consider the following example of a HW module that computes `9 * a` using a left-shift and an addition.
+Note that this is using `verif.ensure_equal` as a standin for `verif.ensure(comb.icmp eq)`.
+The module's output `%z` is the result of a `verif.contract` operation.
+Contracts can be completely ignored by simply passing through their operands, `%1` in this case, to their results.
+This is their normal interpretation outside a formal verification flow, for example for synthesis.
+
+```mlir
+hw.module @Mul9(in %a: i42, out z: i42) {
+  // Compute 9*a as (a<<3)+a.
+  %c3_i42 = hw.constant 3 : i42
+  %0 = comb.shl %a, %c3_i42 : i42
+  %1 = comb.add %a, %0 : i42
+
+  // Contract to check that the circuit actually produces 9*a.
+  %z = verif.contract %1 : i42 {
+    %c9_i42 = hw.constant 9 : i42
+    %a9 = comb.mul %a, %c9_i42 : i42
+    verif.ensure_equal %z, %a9
```

Comment:
You could definitely create an equality assertion implicitly as part of the lowering. This would prevent you from providing symbolic or more general replacement values though. In the carry-save adder compressor example further down, the contract creates symbolic replacement values that are constrained to sum up to a given value. In that example you want to make the contract hide the exact value of the output as an implementation detail.

Reply:
Using the `%z` in the `ensure_equal` instead of the `%1` is important because it allows the original implementation to become unused very easily when the contract is assumed to hold. Otherwise you'd be left with a `verif.assume_equal %1, %a9`, which still uses `%1`. It's not entirely trivial to figure out if that assume could be removed or not :thinking:. For example, the assume could contrain an input further than the contract, and so it might not be safe to get rid of it. Actually making the contract put assurances on the contract output value instead of specifically the implementation fixes that.

Reply:
> This would prevent you from providing symbolic or more general replacement values though.

Good point! In that case you'd want the `ensure_equal` to check the addition of the outputs for equality, which you can't know when inserting them implicitly. But this is also solved by the comment further down about applying the contract to the alternative implementation as well (i.e., the value you return here).

> It's not entirely trivial to figure out if that assume could be removed or not

Right, that's actually a genius way to simplify the implementation here!
If you have the `verif.assume_equal %1, %a9`, the contract already verified that they are the same. So you can just pick whichever you prefer (i.e., you can replace all uses of `%1` with `%a9` during inlining). But, it's an additional step during inlining and thus a bit more work.
Just looking at this IR, it seems like the `ensure` just trivially holds, but you replace `%z` with `%1` when proving the inside. If we'd separate the contract, the 'ensure' would use `%1` always, but you'd inline the alternative representation and thus the DCE is also trivial. On the negative side, you'd need to duplicate the `mul` and constant, but we could also factor out bigger bodies into a func or similar.